pkg/actor/cluster_restart_test.go

```diff
@@ -0,0 +1,215 @@
+				LastTransitionTime: metav1.Now(),
+			},
+			{
+				Type: api.CrdbVersionChecked,
+				Status: metav1.ConditionTrue,
+				LastTransitionTime: metav1.Now(),
+			},
+
+		}
+		require.True(t, cra.Handles(conditions))
+
+	})
+
+	t.Run("Handle func does not pass", func(t *testing.T) {
+
+		// Checked version is false
+		conditions = []api.ClusterCondition{
+			{
+				Type: api.InitializedCondition,
+				Status: metav1.ConditionTrue,
+				LastTransitionTime: metav1.Now(),
+			},
+			{
+				Type: api.CrdbVersionChecked,
+				Status: metav1.ConditionFalse,
+				LastTransitionTime: metav1.Now(),
+			},
+
+		}
+		require.False(t, cra.Handles(conditions))
+
+		// Initialized condition is unknown or missing
+		conditions = []api.ClusterCondition{
+			{
+				Type: api.InitializedCondition,
+				Status: metav1.ConditionUnknown,
+				LastTransitionTime: metav1.Now(),
+			},
+			{
+				Type: api.CrdbVersionChecked,
+				Status: metav1.ConditionTrue,
+				LastTransitionTime: metav1.Now(),
+			},
+
+		}
+		require.False(t, cra.Handles(conditions))
+
+		conditions = []api.ClusterCondition{
+			{
+				Type: api.CrdbVersionChecked,
+				Status: metav1.ConditionTrue,
+				LastTransitionTime: metav1.Now(),
+			},
+
+		}
+		require.False(t, cra.Handles(conditions))
+	})
+
+}
+
+func TestClusterReadyForRestart(t *testing.T) {
+	ssStatus := appsv1.StatefulSetStatus{
+		Replicas: 2,
+		CurrentReplicas: 2,
+	}
+	require.NoError(t, statefulSetReplicasAvailable(&ssStatus))
+
+	ssStatus.Replicas = 0
+	require.Error(t, statefulSetReplicasAvailable(&ssStatus))
+}
+
+func TestFullClusterRestart(t *testing.T) {
+	// Setup fake client
+	builder := fake.NewClientBuilder()
+
+	client := builder.Build()
+
+	cr := newClusterRestart(nil, client, nil).(*clusterRestart)
+	require.NotNil(t, cr)
+	var stsReplicas int32
+	stsReplicas = 3
+	cltSet := fakeclient.NewSimpleClientset()
+
+	sts := createStatefulSet(stsReplicas)
+	cltSet.Tracker().Add(&sts)
+
+	addPodsToStatefulSet(stsReplicas, sts, cltSet)
+
+	cltSet.PrependReactor("*", "*", func(action clienttesting.Action) (handled bool, ret runtime.Object, err error) {
+		tracker := cltSet.Tracker()
+		gvr := action.GetResource()
+		ns := action.GetNamespace()
+		verb := action.GetVerb()
+
+
+		switch verb {
+		case "update":
+			updateAction := action.(clienttesting.UpdateAction)
+			obj := updateAction.GetObject().(*appsv1.StatefulSet)
+			tracker.Update(gvr, obj, ns)
+			return true, obj, nil
+		case "get":
+			getAction := action.(clienttesting.GetAction)
+			obj, err := tracker.Get(gvr, ns, getAction.GetName())
+			if err != nil {
+				return false, nil, err
+			}
+			return true, obj, nil
+		}
+		
+		return false, nil, nil
+	})
+
+	sts.Status.Replicas = stsReplicas
+	sts.Status.ReadyReplicas = stsReplicas
+
+	cltSet.Tracker().Update(schema.GroupVersionResource{
+		Group:    "apps",
+		Version:  "v1",
+		Resource: "statefulset",
+	}, &sts, sts.Namespace)
+	require.NoError(t, cr.fullClusterRestart(context.TODO(), &sts, Log, cltSet))
```

Comment:
@chrisseto this is the call where I'd like to make sure that the restart doesn't complete if all replicas are not ready yet and where I wanted to add in some mocks to add replicas after an initial failure of the check.

Reply:
This is testing to make sure that `fullClusterRestart` does not return an error. These tests are focused on the behaviour of the operator *not* the underlying DB. e.g. If the operator is in this state what are the different actions it takes and then what state is the operator in after that. Any calls to the DB like `HealthChecker` are mocked.

Reply:
👍